src/opencl-poh-verify/cl_poh_verify.cpp

```diff
@@ -2,16 +2,308 @@
+    char temp_string[30];
+    sprintf(temp_string, "%s_%zu_%zu", file_name, index_file, num_elems);
+
+    fp = fopen (temp_string, "w");
+    if (fp == NULL) {
+        fprintf(stderr, "Could not create file %s\n",temp_string);
+        exit(-1);
+    }
+
+    FILE * fp2;
+
+    const char *file_name2 = "test_num_hashes_arr";
+    char temp_string2[25];
+    sprintf(temp_string2, "%s_%zu", file_name2, index_file);
+    fp2 = fopen (temp_string2, "w");
+    if (fp2 == NULL) {
+        fprintf(stderr, "Could not create file %s\n", temp_string2);
+        exit(-2);
+    }
+
+    FILE * fp3;
+
+    const char *file_name3 = "test_num_elems";
+    char temp_string3[25];
+    sprintf(temp_string3, "%s_%zu", file_name3, index_file);
+    fp3 = fopen (temp_string3, "w");
+    if (fp3 == NULL) {
+        fprintf(stderr, "Could not create file %s\n", file_name3);
+        exit(-3);
+    }
+
+    fprintf(fp3, "%lu", num_elems*SHA256_BLOCK_SIZE);
+    fclose(fp3);
+
+    for (size_t i = 0; i < num_elems*SHA256_BLOCK_SIZE; ++i) {
+        fprintf(fp, "%hhu ", hashes[i]);
+    }
+    fclose(fp);
+
+
+    for (size_t i = 0; i < num_elems; ++i) {
+        fprintf(fp2, "%lu ", num_hashes_arr[i]);
+    }
+    fclose(fp2);
+    index_file++;
+
+}
+
+void static inline save_out(uint8_t* hashes,
+    size_t num_elems) {
+
+    FILE * fp;
+
+    const char *file_name = "test_hashes_output_cl";
+    char temp_string[35];
+    sprintf(temp_string, "%s", file_name);
+
+    fp = fopen (temp_string, "w");
+    if (fp == NULL) {
+        fprintf(stderr, "Could not create file %s\n", temp_string);
+        exit(-1);
+    }
+
+    for (size_t i = 0; i < num_elems*SHA256_BLOCK_SIZE; ++i) {
+        fprintf(fp, "%hhu ", hashes[i]);
+    }
+    fclose(fp);
+}
+
 
 extern "C" {
 int poh_verify_many(uint8_t* hashes,
                     const uint64_t* num_hashes_arr,
                     size_t num_elems,
                     uint8_t use_non_default_stream)
 {
-    fprintf(stderr, "poh_verify_many not implemented.");
+    pthread_mutex_lock(&clg_ctx_mutex);
+    bool success = false;
+    //save_input(hashes, num_hashes_arr, num_elems);
+
+    if (initialized == false) {
+        success = cl_check_init();
+        DIE(success == false, "OpenCL could not be init");
+        DIE(cl_crypt_init_locked() == false, "cl_crypt_init_locked failed");
+        initialized = true;
+    } else {
+        fprintf(stderr, "cl_poh_verify_many already initialized\n");
+    }
+
+    gpu_ctx_t *cur_ctx = get_gpu_ctx();
+    pthread_mutex_unlock(&clg_ctx_mutex);
+    
+    size_t nr_bytes_hashes = num_elems * SHA256_BLOCK_SIZE * sizeof(uint8_t);
+    size_t nr_bytes_num_hashes_arr = num_elems * sizeof(uint64_t);
+    setup_gpu_ctx(cur_ctx,
+                    hashes,
+                    num_hashes_arr,
+                    num_elems,
+                    nr_bytes_hashes,
+                    nr_bytes_num_hashes_arr);
+
+    if (num_elems == 0) {
+        return 0;
```

Comment:
Doesn't seem to release the GPU context here.

Reply:
Moved at the begging of the function.